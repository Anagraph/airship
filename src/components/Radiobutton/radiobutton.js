import React, { Children } from 'react';
import styled, { keyframes } from 'styled-components';
import PropTypes from 'prop-types';
import { colors } from '../../constants';
import { rgba } from 'polished';

const radioIn = keyframes`
  from {
    opacity: 0;
    transform: scale3d(0.3, 0.3, 0.3);
    transform-origin: 50% 50%;
  }
  to {
    opacity: 1;
    transform: scale3d(1, 1, 1);
  }
`;

const StyleRadio = styled.div`
  display: flex;
  align-items: center;

  .Radio {
    position: relative;
  }

  .Radio-decoration {
    pointer-events: none;
    width: 16px;
    height: 16px;
    border: 1px solid ${rgba(colors.type01, 0.16)};
    border-radius: 50%;
    position: relative;
    display: block;
  }

  .Radio-decoration:after {
    content: '';
    display: block;
    border-radius: 50%;
    position: absolute;
    top: calc(50% - 3px);
    left: calc(50% - 3px);
    width: 6px;
    height: 6px;
  }

  .Radio-input {
    appearance: none;
    background: none;
    width: 16px;
    height: 16px;
    position: absolute;
    top: 0;
    left: 0;
    cursor: pointer;
    margin: 0;
    padding: 0;
    border-radius: 50%;
  }

  .Radio-input:focus {
    border: 2px solid ${colors.brand01};
    outline: none;
  }

  .Radio-input:hover + .Radio-decoration {
    border: 1px solid ${colors.brand03};
  }

  // .Radio-input:hover + .Radio-decoration:after {
  //   background: ${colors.brand01};
  //   width: 6px;
  //   height: 6px;
  //   opacity: 0;
  //   animation: ${radioIn} 300ms;
  //   animation-fill-mode: forwards;
  // }

  // .Radio-input:checked:hover + .Radio-decoration {
  //   animation: none;
  // }

  // .Radio-input:checked:hover + .Radio-decoration:after {
  //   background: ${colors.brand01};
  //   opacity: 1;
  //   animation: initial;
  // }

  .Radio-input:checked + .Radio-decoration {
    border: 1px solid ${colors.brand01};
  }

  // .Radio-input:checked + .Radio-decoration:before {
  //   content: '';
  //   display: block;
  //   width: 16px;
  //   height: 16px;
  //   opacity: 1;
  //   background: ${colors.white};
  //   // animation: ${radioIn} 600ms;
  //   // animation-fill-mode: forwards;
  //   position: absolute;
  //   top: calc(50% - 8px);
  //   left: calc(50% - 8px);
  //   border-radius: 50%;
  // }

  .Radio-input:checked + .Radio-decoration::after {
    opacity: 1;
    background: ${colors.brand01};
    animation: ${radioIn} 300ms;
    animation-fill-mode: forwards;
  }

  .Radio-label {
    margin-left: 4px;
  }
`;

const StyledGroup = styled.div``;

const Group = ({ children, name, onChange, as = 'ul' }) => {
  const Wrapper = as !== 'div' ? StyledGroup.withComponent(as) : StyledGroup;
  return (
    <Wrapper>
      {Children.map(children, (child, index) => {
        return React.cloneElement(child, {
          name: name,
          onChange: onChange,
          key: `radio${index}`,
          as: as === 'ul' ? 'li' : 'div'
        });
      })}
    </Wrapper>
  );
};

Group.propTypes = {
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  as: PropTypes.oneOf(['div', 'ul', 'span'])
};

const RadioButton = ({ children, name, value, as, onChange }) => {
  const Wrapper = as !== 'div' ? StyleRadio.withComponent(as) : StyleRadio;
  return (
    <Wrapper>
      <div className="Radio">
        <input
          className="Radio-input"
          type="radio"
          name={name}
          value={value}
          onChange={onChange}
        />
        <span className="Radio-decoration" />
      </div>
      {children && <div className="Radio-label">{children}</div>}
    </Wrapper>
  );
};

RadioButton.Group = Group;

RadioButton.defaultProps = {
  name: '',
  as: 'div',
  onChange: (e) => {}
};

RadioButton.propTypes = {
  name: PropTypes.string,
  value: PropTypes.string.isRequired,
  as: PropTypes.oneOf(['div', 'li', 'span']),
  onChange: PropTypes.func
};

export default RadioButton;
