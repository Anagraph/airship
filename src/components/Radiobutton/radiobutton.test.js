import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import Radiobutton from './radiobutton';

describe('render', () => {
  it('renders without crashing', () => {
    const component = renderer.create(<Radiobutton value="wadus" />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with label', () => {
    const component = renderer.create(
      <Radiobutton value="wadus">Wadus</Radiobutton>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders radio button group', () => {
    const component = renderer.create(
      <Radiobutton.Group name="wadus">
        <Radiobutton value="hola">Hola</Radiobutton>
        <Radiobutton value="mundo">Mundo</Radiobutton>
      </Radiobutton.Group>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('onChange', () => {
    const spy = jest.fn();
    const component = mount(
      <Radiobutton.Group name="wadus" onChange={spy}>
        <Radiobutton value="hola">Hola</Radiobutton>
        <Radiobutton value="mundo">Mundo</Radiobutton>
      </Radiobutton.Group>
    );

    component
      .find('input')
      .first()
      .simulate('change');
    expect(spy).toHaveBeenCalled();
  });

  it('selected', () => {
    const component = mount(
      <Radiobutton.Group name="wadus" selected="mundo">
        <Radiobutton value="hola">Hola</Radiobutton>
        <Radiobutton value="mundo">Mundo</Radiobutton>
      </Radiobutton.Group>
    );

    const radio = component.find('input').at(1);
    expect(radio.props().checked).toBe(true);
  });
});
