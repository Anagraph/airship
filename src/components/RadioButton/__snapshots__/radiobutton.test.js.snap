// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render renders radio button group 1`] = `
.c0 > label {
  margin-right: 8px;
}

.c0 > label:last-child {
  margin-right: 0;
}

.c6 {
  margin-left: 4px;
  font: 400 12px/20px 'Roboto';
}

.c2 {
  position: relative;
}

.c5 {
  pointer-events: none;
  width: 16px;
  height: 16px;
  border: 1px solid rgba(44,44,44,0.16);
  border-radius: 50%;
  position: relative;
  display: block;
}

.c5:after {
  content: '';
  display: block;
  border-radius: 50%;
  position: absolute;
  top: calc(50% - 3px);
  left: calc(50% - 3px);
  width: 6px;
  height: 6px;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
}

.c3 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: none;
  width: 16px;
  height: 16px;
  position: absolute;
  top: 0;
  left: 0;
  cursor: pointer;
  margin: 0;
  padding: 0;
  border-radius: 50%;
}

.c3:focus {
  outline: none;
}

.c3:hover + .c4 {
  border: 1px solid #47DB99;
}

.c3:checked + .c4 {
  border: 1px solid #1785FB;
}

.c3:checked:disabled + .c4 {
  border: 1px solid rgba(44,44,44,0.16);
}

.c3:checked + .c4::after {
  opacity: 1;
  background: #1785FB;
  -webkit-animation: jMKEOZ 300ms;
  animation: jMKEOZ 300ms;
  -webkit-animation-fill-mode: forwards;
  animation-fill-mode: forwards;
}

.c3:checked:disabled + .c4::after {
  background: rgba(44,44,44,0.16);
}

.c3:disabled:hover {
  cursor: not-allowed;
}

.c3:disabled + .c4 {
  background-color: #E2E6E3;
}

.c3:disabled:hover + .c4 {
  border: 1px solid rgba(44,44,44,0.16);
}

<div
  className="c0"
>
  <label
    className="c1"
  >
    <div
      className="c2"
    >
      <input
        checked={false}
        className="c3"
        disabled={false}
        name="wadus"
        onChange={[Function]}
        type="radio"
        value="hola"
      />
      <span
        className="c4 c5"
      />
    </div>
    <div
      className="c6"
    >
      Hola
    </div>
  </label>
  <label
    className="c1"
  >
    <div
      className="c2"
    >
      <input
        checked={false}
        className="c3"
        disabled={false}
        name="wadus"
        onChange={[Function]}
        type="radio"
        value="mundo"
      />
      <span
        className="c4 c5"
      />
    </div>
    <div
      className="c6"
    >
      Mundo
    </div>
  </label>
</div>
`;

exports[`render renders with label 1`] = `
.c5 {
  margin-left: 4px;
  font: 400 12px/20px 'Roboto';
}

.c1 {
  position: relative;
}

.c4 {
  pointer-events: none;
  width: 16px;
  height: 16px;
  border: 1px solid rgba(44,44,44,0.16);
  border-radius: 50%;
  position: relative;
  display: block;
}

.c4:after {
  content: '';
  display: block;
  border-radius: 50%;
  position: absolute;
  top: calc(50% - 3px);
  left: calc(50% - 3px);
  width: 6px;
  height: 6px;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
}

.c2 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: none;
  width: 16px;
  height: 16px;
  position: absolute;
  top: 0;
  left: 0;
  cursor: pointer;
  margin: 0;
  padding: 0;
  border-radius: 50%;
}

.c2:focus {
  outline: none;
}

.c2:hover + .c3 {
  border: 1px solid #47DB99;
}

.c2:checked + .c3 {
  border: 1px solid #1785FB;
}

.c2:checked:disabled + .c3 {
  border: 1px solid rgba(44,44,44,0.16);
}

.c2:checked + .c3::after {
  opacity: 1;
  background: #1785FB;
  -webkit-animation: jMKEOZ 300ms;
  animation: jMKEOZ 300ms;
  -webkit-animation-fill-mode: forwards;
  animation-fill-mode: forwards;
}

.c2:checked:disabled + .c3::after {
  background: rgba(44,44,44,0.16);
}

.c2:disabled:hover {
  cursor: not-allowed;
}

.c2:disabled + .c3 {
  background-color: #E2E6E3;
}

.c2:disabled:hover + .c3 {
  border: 1px solid rgba(44,44,44,0.16);
}

<label
  className="c0"
>
  <div
    className="c1"
  >
    <input
      checked={false}
      className="c2"
      disabled={false}
      name=""
      onChange={[Function]}
      type="radio"
      value="wadus"
    />
    <span
      className="c3 c4"
    />
  </div>
  <div
    className="c5"
  >
    Wadus
  </div>
</label>
`;

exports[`render renders without crashing 1`] = `
.c1 {
  position: relative;
}

.c4 {
  pointer-events: none;
  width: 16px;
  height: 16px;
  border: 1px solid rgba(44,44,44,0.16);
  border-radius: 50%;
  position: relative;
  display: block;
}

.c4:after {
  content: '';
  display: block;
  border-radius: 50%;
  position: absolute;
  top: calc(50% - 3px);
  left: calc(50% - 3px);
  width: 6px;
  height: 6px;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
}

.c2 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: none;
  width: 16px;
  height: 16px;
  position: absolute;
  top: 0;
  left: 0;
  cursor: pointer;
  margin: 0;
  padding: 0;
  border-radius: 50%;
}

.c2:focus {
  outline: none;
}

.c2:hover + .c3 {
  border: 1px solid #47DB99;
}

.c2:checked + .c3 {
  border: 1px solid #1785FB;
}

.c2:checked:disabled + .c3 {
  border: 1px solid rgba(44,44,44,0.16);
}

.c2:checked + .c3::after {
  opacity: 1;
  background: #1785FB;
  -webkit-animation: jMKEOZ 300ms;
  animation: jMKEOZ 300ms;
  -webkit-animation-fill-mode: forwards;
  animation-fill-mode: forwards;
}

.c2:checked:disabled + .c3::after {
  background: rgba(44,44,44,0.16);
}

.c2:disabled:hover {
  cursor: not-allowed;
}

.c2:disabled + .c3 {
  background-color: #E2E6E3;
}

.c2:disabled:hover + .c3 {
  border: 1px solid rgba(44,44,44,0.16);
}

<label
  className="c0"
>
  <div
    className="c1"
  >
    <input
      checked={false}
      className="c2"
      disabled={false}
      name=""
      onChange={[Function]}
      type="radio"
      value="wadus"
    />
    <span
      className="c3 c4"
    />
  </div>
</label>
`;
